src/interface.jl

```diff
@@ -4,14 +4,16 @@ struct Leaf{R,S}
   state::S
 end
 
-function setup(rule, x)
+function setup(rule, x; seen = Base.IdSet())
   if isnumeric(x)
+    x in seen && throw(ArgumentError("Optimisers.jl does not at present handle tied weights, sorry."))
```

Comment:
```suggestion
    iscachesafe(x) && x in seen && throw(ArgumentError("Optimisers.jl does not at present handle tied weights, sorry."))
```
Where:
```julia
iscachesafe(x) = !isbits(x) && ismutable(x)
iscachesafe(::Union{String,Symbol}) = false
```
To prevent false positives on value types.

Reply:
This should probably be in functors and used liberally whenever a cache is in play.

Reply:
Oh I forgot about strings. I don't think ismutable tells you much, as an immutable wrapper can hold a mutable object.

Reply:
No it's fine. Strings can't be `isnumeric`. And I test `isbits` before pushing into the cache.

Reply:
I was backfilling this implementation from work on functors and missed the `isnumeric` guard above. Would recommend checking for `isbits` first before doing anything requiring `objectid`, as `objectid` is significantly slower for value types (involves a ccall and hashing):
```julia-repl
julia> x = ntuple(identity, 100);

julia> @btime objectid($x)
  99.999 ns (0 allocations: 0 bytes)
0x31d9b387e6e6f021

julia> @btime isbits($x)
  1.152 ns (0 allocations: 0 bytes)
```
How about:
```julia
if !isbits(x)
  x in seen && throw(ArgumentError("Optimisers.jl does not at present handle tied weights, sorry."))
  push!(seen, x)
end
```

Reply:
Agree this is faster, but... Zygote doesn't really work with SArrays right now, so it's just numbers, and this runs exactly once while Functors's `fmap` might run several times per iteration... so my vote is that this doesn't really merit much optimisation yet.

Reply:
I thought Zygote could project back out to SArrays now? But yes, `fmap`s cache usage is a better optimization target and may have already caused performance issues (there's an issue about slow `fcollect` somewhere).

Reply:
It tries, and wraps many in SizedArray. But IIRC many things fail, e.g. broadcasting uses `sum(SA[1 2; 3 4]; dims=(1,3))` (ndims+1 as a hack to get type-stable dims).